### ISSUE TITLE
The log `FileHandler` did not write the log to the file

### ISSUE THREAD

**Issue by yi-huan** [denoland/std#3676]
**Describe the bug**
I used logs [example](https://deno.land/std@0.202.0/log/mod.ts#example_0),
If `Deno.serve` is used, `FileHandler` will not be able to write to the file.

**Steps to Reproduce**

`main.ts`

```typescript
import * as log from 'https://deno.land/std@0.203.0/log/mod.ts';

log.setup({
  handlers: {
    console: new log.handlers.ConsoleHandler('DEBUG'),

    file: new log.handlers.FileHandler('WARNING', {
      filename: './log.txt',
      formatter: '{levelName} {msg}',
    }),
  },

  loggers: {
    default: {
      level: 'DEBUG',
      handlers: ['console', 'file'],
    },
  },
});

const logger = log.getLogger();
logger.debug('Display on the console');
logger.warning('Display on the console and write to file');

await Deno.serve((_req) => {
  logger.warning(Date.now());
  return new Response('Hello, World!');
}).finished;

```

use:
```shell
deno run -A main.ts
```

**Expected behavior**

'Display on the console and write to file' will be written to the file `log.txt`.

I only saw the file created, There is no content inside.

**Environment**

- OS: [Windows 11 专业版 22H2 22621.2134]
- deno version: 1.37.1
- std version: 0.203.0

**Comment by yi-huan** [comment]
When there is no `Deno.serve`,
```typescript
// await Deno.serve((_req) => {
//   logger.warning(Date.now());
//   return new Response('Hello, World!');
// }).finished;
```
 **it works as expected.**

Is there something causing the conflict?

**Comment by markthree** [comment]
> When there is no `Deno.serve`,
> 
> ```ts
> // await Deno.serve((_req) => {
> //   logger.warning(Date.now());
> //   return new Response('Hello, World!');
> // }).finished;
> ```
> 
> **it works as expected.**
> 
> Is there something causing the conflict?

There is an internal caching mechanism, so you need to flush it manually

```ts
import * as log from "https://deno.land/std@0.203.0/log/mod.ts";

const fileHandler = new log.handlers.FileHandler("WARNING", {
  filename: "./log.txt",
  formatter: "{levelName} {msg}",
});

log.setup({
  handlers: {
    console: new log.handlers.ConsoleHandler("DEBUG"),
    file: fileHandler,
  },

  loggers: {
    default: {
      level: "DEBUG",
      handlers: ["console", "file"],
    },
  },
});

const logger = log.getLogger();
logger.debug("Display on the console");
logger.warning("Display on the console and write to file");

await Deno.serve((_req) => {
  logger.warning(Date.now());
  fileHandler.flush(); // Manual flush
  return new Response("Hello, World!");
}).finished;
```

**Comment by yi-huan** [comment]
thanks @markthree .
Without further explanation in the documentation, this is difficult to use.

If I don't flush, when will he write?

Also, if I don't refresh, I need to refresh when I need to shut down and restart, I don't know if this design is friendly.

**Comment by markthree** [comment]
@yi-huan Yes, I'm trying to figure it out, waiting for deno std to reach version 1.0.0 which should have more detailed documentation。

Currently it only triggers flush when the cache limit is reached, or when the process is unloaded.

Of course it can be triggered manually。

**Comment by yi-huan** [comment]
Okay, thank you.  I hope there will be other ways to write the in-memory logs to a file when exiting abnormally.

**Comment by markthree** [comment]
@yi-huan you can use [easy_std/src/process](https://github.com/markthree/easy_std/blob/main/src/process.ts#L60)

for example ↓
```ts
import * as log from "https://deno.land/std@0.203.0/log/mod.ts";
import { gracefulShutdown } from "https://deno.land/x/easy_std@v0.5.2/src/process.ts";

const fileHandler = new log.handlers.FileHandler("WARNING", {
  filename: "./log.txt",
  formatter: "{levelName} {msg}",
});

// The incoming function will be executed regardless of the reason for the process to exit
gracefulShutdown(() => fileHandler.flush());

log.setup({
  handlers: {
    console: new log.handlers.ConsoleHandler("DEBUG"),

    file: fileHandler,
  },

  loggers: {
    default: {
      level: "DEBUG",
      handlers: ["console", "file"],
    },
  },
});

const logger = log.getLogger();
logger.debug("Display on the console");
logger.warning("Display on the console and write to file");

await Deno.serve((_req) => {
  logger.warning(Date.now());
  return new Response("Hello, World!");
}).finished;
```

**Comment by yi-huan** [comment]
@markthree thanks, It's easy to use.

**Comment by fushihara** [comment]
In the current source code, the buffer size seems to be fixed at 4096 bytes in PAGE_SIZE.
My program has a low load and logs only one case every few seconds, so I would like to reduce the buffer size.
I see no reason why this value cannot be specified in the constructor, so I hope it will be changeable.

But I can't wait, so I created my own inherited class.
I ran it only once, but I don't think there is any problem.

```ts
class RotatingFileHandler2 extends RotatingFileHandler {
  constructor(levelName: ConstructorParameters<typeof RotatingFileHandler>[0], options: ConstructorParameters<typeof RotatingFileHandler>[1]) {
    super(levelName, options);
    this._buf = new Uint8Array(1);
  }
}
```

@markthree 
I understand that this is a unique situation,
but when I am debugging with vscode and force quit the deno with the stop button, this logic does not seem to work.

**Comment by iuioiua** [comment]
> In the current source code, the buffer size seems to be fixed at 4096 bytes in PAGE_SIZE. My program has a low load and logs only one case every few seconds, so I would like to reduce the buffer size. I see no reason why this value cannot be specified in the constructor, so I hope it will be changeable.
> 
> But I can't wait, so I created my own inherited class. I ran it only once, but I don't think there is any problem.
> 
> ```ts
> class RotatingFileHandler2 extends RotatingFileHandler {
>   constructor(levelName: ConstructorParameters<typeof RotatingFileHandler>[0], options: ConstructorParameters<typeof RotatingFileHandler>[1]) {
>     super(levelName, options);
>     this._buf = new Uint8Array(1);
>   }
> }
> ```
> 
> @markthree I understand that this is a unique situation, but when I am debugging with vscode and force quit the deno with the stop button, this logic does not seem to work.

I've created #4674 for this suggestion, as it's a separate issue.

**Comment by yi-huan** [comment]
> In the current source code, the buffer size seems to be fixed at 4096 bytes in PAGE_SIZE. My program has a low load and logs only one case every few seconds, so I would like to reduce the buffer size. I see no reason why this value cannot be specified in the constructor, so I hope it will be changeable.
> 
> But I can't wait, so I created my own inherited class. I ran it only once, but I don't think there is any problem.
> 
> ```ts
> class RotatingFileHandler2 extends RotatingFileHandler {
>   constructor(levelName: ConstructorParameters<typeof RotatingFileHandler>[0], options: ConstructorParameters<typeof RotatingFileHandler>[1]) {
>     super(levelName, options);
>     this._buf = new Uint8Array(1);
>   }
> }
> ```
> 
> @markthree I understand that this is a unique situation, but when I am debugging with vscode and force quit the deno with the stop button, this logic does not seem to work.

Abnormal exits should not be executed at all times. It may only be possible to process it outside of the program.

**Comment by cristian-spiescu** [comment]
> @yi-huan you can use [easy_std/src/process/gracefulShutdown](https://github.com/markthree/easy_std/blob/main/src/process.ts#L60)

I confirm that this solution works. However, in my opinion:

1/ such behavior should be implemented by default in the file handler. I.e. app stops abruptly => the log until that moment should exist on the disk. As a temp measure, until implementing this (properly), this gotcha should be present in the doc in ALL CAPS. In my case, I learned it the hard way. The app was crashing. No logs were recorded. And the investigation led me here.

2/ w/ the "properly" word from above I mean: `gracefulShutdown()` is a solution, but it is not bullet proof. What if the JS VM crashes? In this case no more graceful shutdown => (maybe precious) unwritten logs won't be written on disk.

I come from the Java world. Mature logging frameworks work by default cf. 2/. Some have an optional mode "async", faster, but w/ the drawback that in case of program crash, some entries may be lost.

**Comment by markthree** [comment]
@cristian-spiescu 

Yes, `gracefulShutdown` only cares about exiting at the js level, if it's an unknown error, e.g. vm crash, OS crash, then it will fail.

**Comment by yi-huan** [comment]
@cristian-spiescu Yes, it is possible to use this `gracefulShutdown` to log when manually closing or starting a program, but I think it should be implemented by std's log. 
In addition, If it's a program crash, it should only be recorded by the deno implementation.

**Comment by kt3k** [comment]
We removed std/log in https://github.com/denoland/std/pull/6819

See the discussions in https://github.com/denoland/std/issues/6124
